SWC-6610 - Add ColumnModelsEditorV2Widget

nickgros opened the pull request:
- Add ColumnModelsEditorV2Widget which will be shown when editing table columns in experimental mode
 - Update ColumnModelsWidget and CreateTableViewWizardStep2 to utilize new editor
 - Add JSONEntityUtils.toJsInteropCompatibleObject to simplify converting a JSONEntity to a JsInterop-compatible native JavaScript object
 - Update JsInterop definitions for React to support handling component refs

nickgros committed:
SWC-6610 - Add ColumnModelsEditorV2Widget
 - Add ColumnModelsEditorV2Widget which will be shown when editing table columns in experimental mode
 - Update ColumnModelsWidget and CreateTableViewWizardStep2 to utilize new editor
 - Add JSONEntityUtils.toJsInteropCompatibleObject to simplify converting a JSONEntity to a JsInterop-compatible native JavaScript object
 - Update JsInterop definitions for React to support handling component refs

nickgros commented on src/main/java/org/sagebionetworks/web/client/jsinterop/AccessRequirementListProps.java:
```diff
@@ -33,7 +33,7 @@ public static AccessRequirementListProps create(
       accessRequirements
         .stream()
-        .map(o -> JSON.parse(o.toJSONString()))
+        .map(JSONEntityUtils::toJsInteropCompatibleObject)
```
The new TableColumnSchemaForm component also serialize objects in this way, so add and use a common utility.

nickgros commented on src/main/java/org/sagebionetworks/web/client/jsinterop/React.java:
```diff
@@ -25,6 +25,8 @@ public static native <P extends ReactComponentProps> ReactNode createElement(
     ReactNode[] children
   );
 
+  public static native <T> T createRef();
+
```
Expose `createRef`

nickgros committed:
SWC-6610 - Stronger typing for the ref object

nickgros committed:
SWC-6610 - Fix tests

nickgros committed:
SWC-6610 - Bump synapse-react-client to 3.1.64

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/accessrequirements/ManagedACTAccessRequirementWidgetViewImpl.java:
```diff
@@ -363,38 +361,29 @@ public void showRequestAccessModal(
     ManagedACTAccessRequirement accessRequirement,
     @Nullable RestrictableObjectDescriptor targetSubject
   ) {
-    try {
```
Refactored so JSON serialization happens in the new utility, so the `try`/`catch` could be removed.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/table/modal/fileview/CreateTableViewWizardStep2.java:
```diff
@@ -118,30 +123,45 @@ public void configure(Table entity, TableType tableType) {
+    } else {
+      editor.configure(tableType, new ArrayList<>());
+      boolean isView = !TableType.table.equals(tableType);
+      this.editor.setAddDefaultColumnsButtonVisible(isView);
+      this.editor.setAddAnnotationColumnsButtonVisible(isView);
```
These visibility conditions are handled internally by the React component, so the methods on the new widget are not 1:1 with the methods on the old widget.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/table/v2/schema/ColumnModelsEditorV2Widget.java:
```diff
@@ -0,0 +1,40 @@
+import org.sagebionetworks.repo.model.table.ColumnModel;
+import org.sagebionetworks.repo.model.table.ViewScope;
+
+public class ColumnModelsEditorV2Widget implements IsWidget {
```
I considered extracting a common interface between ColumnModelsEditorWidget (v1) and ColumnModelsEditorV2Widget, which would have simplified initial changes to the classes that use the editor. It would have been a fairly complicated refactor because

1. ColumnModelsEditorWidget (v1) also implements its view's presenter interface, which would be inappropriate to implement in v2. I'm not sure if the view is the only class to access the presenter methods, so it may have been tough to extract a new non-presenter interface.
2. The v2 implementation handles more logic internally, so it wouldn't have as many methods, so we would have ended up needing to implement many no-op methods. 
3. I don't immediately know if/how to use Guice to accomplish conditional dependency injection based on the experimental mode cookie. The logic could be in the constructor would work but it's not simple to reason with since some of our widgets are persistent and just get reconfigured while others are regularly destroyed and re-created


I ultimately decided that it wasn't worth it for temporary code.

jay-hodgson replied:
Totally agree, no need to share a common interface.  In general, I think the more logic that's in new implementation, the less complicated the interface will be.

jay-hodgson replied:
The separation will make deletion of v1 that much easier.

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/table/v2/schema/ColumnModelsEditorV2WidgetViewImpl.java:
```diff
@@ -0,0 +1,100 @@
+          adapterFactory.createNew(JSON.stringify(o))
+        );
+        asList.add(cm);
+      } catch (JSONObjectAdapterException e) {
+        e.printStackTrace();
+      }
```
I could add an alert container to the widget, or propagate this up, but I don't think this is common. Thoughts?

jay-hodgson replied:
I don't expect this parsing error to ever happen, I think sending it to the console is ok

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/table/v2/schema/ColumnModelsEditorV2WidgetViewImpl.java:
```diff
@@ -0,0 +1,100 @@
+    ViewScope viewScope,
+    List<ColumnModel> startingModels
+  ) {
+    componentRef = React.createRef();
```
I was so happy to find out that this worked to give us a stable component ref

jay-hodgson replied:
I'm a bit surprised that this worked!

nickgros commented on src/main/java/org/sagebionetworks/web/client/widget/table/v2/schema/ColumnModelsWidget.java:
```diff
@@ -171,18 +166,28 @@ public void getPossibleColumnModelsForViewScope(String nextPageToken) {
       scopeIds = new ArrayList<>();
       for (EntityRef item : ((Dataset) entity).getItems()) {
-        scopeIds.add(item.getEntityId());
+        scopeIds.add(
+          item.getEntityId() + "." + item.getVersionNumber().toString()
+        );
       }
       scope.setViewEntityType(ViewEntityType.dataset);
     } else if (entity instanceof DatasetCollection) {
       scopeIds = new ArrayList<>();
       for (EntityRef item : ((DatasetCollection) entity).getItems()) {
         scopeIds.add(item.getEntityId());
+        scopeIds.add(
+          item.getEntityId() + "." + item.getVersionNumber().toString()
+        );
```
Fixes SWC-6368

nickgros commented on src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java:
```diff
@@ -1,6 +1,7 @@
 package org.sagebionetworks.web.server.servlet;
 
 import au.com.bytecode.opencsv.CSVReader;
+import com.google.gwt.core.client.GWT;
```
```suggestion
```

nickgros commented on src/test/java/org/sagebionetworks/web/unitclient/widget/table/modal/fileview/CreateTableViewWizardStep2Test.java:
```diff
@@ -153,9 +167,12 @@ public class CreateTableViewWizardStep2Test {
-    when(mockTableEntity.getId()).thenReturn(ENTITY_ID);
-    when(mockEntityView.getId()).thenReturn(ENTITY_ID);
-    when(mockSubmissionView.getId()).thenReturn(ENTITY_ID);
+    mockTableEntity = new TableEntity();
+    mockEntityView = new EntityView();
+    mockSubmissionView = new SubmissionView();
+    mockTableEntity.setId(ENTITY_ID);
+    mockEntityView.setId(ENTITY_ID);
+    mockSubmissionView.setId(ENTITY_ID);
```
We are now passing these to EntityTypeUtils, which checks the class to give us an `EntityType`. That check doesn't work with mocks, so just use "real" data objects.

nickgros committed:
Remove unnecessary import

jay-hodgson commented on src/main/java/org/sagebionetworks/web/client/widget/table/modal/fileview/CreateTableViewWizardStep2.java:
```diff
@@ -118,30 +123,45 @@ public void configure(Table entity, TableType tableType) {
     List<String> scopeIds = null;
-    if (entity instanceof EntityView) {
+    if (entity instanceof TableEntity) {
+      return new ViewScope();
```
A TableEntity has a ViewScope?

nickgros replied:
I think I hit an NPE but the right thing to do would be to return null and explicitly handle the null case, thanks for catching this.

jay-hodgson commented on src/main/resources/org/sagebionetworks/web/client/widget/table/v2/schema/ColumnModelsEditorV2WidgetViewImpl.ui.xml:
```diff
@@ -0,0 +1,10 @@
+  xmlns:b="urn:import:org.gwtbootstrap3.client.ui"
+  xmlns:b.html="urn:import:org.gwtbootstrap3.client.ui.html"
+  xmlns:t="urn:import:org.sagebionetworks.web.client.view.bootstrap.table"
+  xmlns:w="urn:import:org.sagebionetworks.web.client.widget"
```
Could clean these imports (I think you just need w)

jay-hodgson commented on src/test/java/org/sagebionetworks/web/unitclient/widget/table/modal/fileview/CreateTableViewWizardStep2Test.java:
```diff
@@ -153,9 +167,12 @@ public class CreateTableViewWizardStep2Test {
-    when(mockSubmissionView.getId()).thenReturn(ENTITY_ID);
+    mockTableEntity = new TableEntity();
+    mockEntityView = new EntityView();
+    mockSubmissionView = new SubmissionView();
```
These could be renamed to be more appropriate (tableEntityStub for example)

jay-hodgson commented on pom.xml:
```diff
@@ -19,7 +19,7 @@
       ${project.build.directory}/${project.build.finalName}
     </webappDirectory>
-    <synapse.version>462.0</synapse.version>
+    <synapse.version>475.0</synapse.version>
```
Thank you for updating the synapse version as well!

nickgros replied:
It was necessary to properly serialize the JSON Subcolumns!

nickgros committed:
SWC-6610 - Changes from code review
 - Table should have null view scope, handle null case in props serializer
 - Remove extra imports in XML
 - Rename variables from mock -> stub in test
